### PR TITLE
Parser for TZ

### PR DESCRIPTION
**Pull request by Yuri6037**
This provides an implementation of the TZ POSIX environment variable specification for timezone information.

This exposes a new API function `time_tz::parse_tz::parse(input: &str)` which returns an owned type implementing the `TimeZone` trait.

This currently needs a lot of testing to ensure the hacks works and the many "unwrap" also passes.

Link to the spec: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap08.html#tag_08_03

### CODE REVIEW

File: src/parse_tz.rs [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+        self.inner.get_offset_utc(date_time)
+    }
+
+    fn name(&self) -> &str {
+        self.inner.name()
+    }
+}
+
+pub fn parse(input: &str) -> Result<ParsedTz, Error> {
+    let (_, inner) = entry(input).map_err(|v| match v {
```
Comment (MiSawa) [inline src/parse_tz.rs:426]: I think we need to ensure there's no string follows this by `terminated(entry, eof)(input)` (or maybe put it in the `entry`), or otherwise `STD+1#whatever` for example would be parsed as `STD+1`.

**Review by MiSawa — COMMENTED** [review]
Brilliant! I noticed some parts that doesn't seem to match the specification, so commented them along with some nitpicks.

File: src/parse_tz.rs [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+        time_component,
+        opt(time_component_opt),
+        opt(time_component_opt),
+    ))(input)?;
+    Ok((input, Time { hh, mm, ss }))
+}
+
+fn offset(input: &str) -> IResult<&str, Offset> {
+    let (input, (sign, time)) = tuple((opt(sign), time))(input)?;
+    let positive = sign.map(|v| v == '+').unwrap_or(true);
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:243]: In that case that's a real problem because that would be dependent over OffsetDateTime::now_local which returns local time and also would be really strange because that would mean you must call localtime which uses TZ in order to compute TZ.

The only other way I can think would be to use time-tz to compute localtime using current system timezone but that is inaccurate for windows systems... This also means that it becomes impossible to use parse-tz independently of the system feature and also parse-tz cannot be used as a timezone then...

Othe possibility is we're not standard compliant and assume that the date_time given as input is the localtime which in reality is most likely not the case...

[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:243]: Actually a little math saves the day:

`local+offset=utc`
`<=> local=utc+(-offset)`

so that basically means you just take the opposite of the offset. Well yeah easy fix

File: src/parse_tz.rs [hunk 3]
```diff
@@ -0,0 +1,487 @@
...
+                                let tz_offset = timezone.get_offset_utc(date_time);
+                                if tz_offset.is_dst() {
+                                    ParsedTzOffset::Expanded(dst.name, dst_offset, true)
+                                } else {
+                                    ParsedTzOffset::Expanded(std.name, std_offset, false)
+                                }
+                            }
+                            Some(rule) => {
+                                let start_date = rule.start.0.to_date(date_time.year());
+                                let end_date = rule.end.0.to_date(date_time.year());
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:370]: we could convert date_time to UTC this way the year is always according to UTC time.

File: src/parse_tz.rs [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+        Ok(())
+    }
+}
+
+pub enum RangeError {
+    Time,
+    Date,
+}
+
+pub enum Error<'a> {
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:191]: oops yeah forgot the impls (that's why it's named experimental :)). Thanks for noticing

File: src/parse_tz.rs [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+                // contains 28 days instead of 29 which matches the POSIX spec.
+                let date = time::Date::from_ordinal_date(2021, *n).unwrap();
+                // Not sure if that will work in all cases though...
+                time::Date::from_calendar_date(year, date.month(), date.day()).unwrap()
+            }
+            // ComponentRange errors should be prevented by is_valid_range.
+            Date::N(n) => time::Date::from_ordinal_date(year, *n + 1).unwrap(),
+            Date::M { m, n, d } => {
+                // One more hack: here w're trying to match Date::from_iso_week_date.
+                let week = m * n;
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:99]: yeah I wasn't particularly sure about this one (that's why I've added "hack")

File: src/parse_tz.rs [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+        match self {
+            Date::J(n) => {
+                // Hack: the basic idea is 2021 was not a leap year so february only
+                // contains 28 days instead of 29 which matches the POSIX spec.
+                let date = time::Date::from_ordinal_date(2021, *n).unwrap();
+                // Not sure if that will work in all cases though...
+                time::Date::from_calendar_date(year, date.month(), date.day()).unwrap()
+            }
+            // ComponentRange errors should be prevented by is_valid_range.
+            Date::N(n) => time::Date::from_ordinal_date(year, *n + 1).unwrap(),
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:96]: oh yeah pretty much all 3 cases can error. Well that's gonna be a problem considering that the TimeZone trait cannot pass errors and of course changing that signature is a breaking change, I was hoping to distribute this as a non-breaking change.

File: src/parse_tz.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs:370]: or better yet: pre-compute the entire OffsetDateTime using UTC as base, then we assume date_time is of whatever offset/whatever local time the user wants.

File: src/parse_tz.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: This is now fixed in a non-breaking way with latest commit.

File: src/parse_tz.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: This is now fixed with latest commit.

File: src/parse_tz.rs [hunk 7]
```diff
@@ -0,0 +1,487 @@
...
+            ParsedTz1::Expanded((std, dst)) => {
+                let std_offset = UtcOffset::from_whole_seconds(std.offset.to_seconds()).unwrap();
+                match dst {
+                    None => {
+                        //Easy case
+                        ParsedTzOffset::Expanded(std.name, std_offset, false)
+                    }
+                    Some(dst) => {
+                        // If the offset is undefined then assume +1 as specified by POSIX.
+                        let offset = dst.offset.as_ref().map(|v| v.to_seconds()).unwrap_or(3600);
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Yep fixed in latest commit.

File: src/parse_tz.rs [hunk 8]
```diff
@@ -0,0 +1,487 @@
...
+                                    .as_ref()
+                                    .map(|v| v.to_time())
+                                    .unwrap_or(default);
+                                let end_time =
+                                    rule.end.1.as_ref().map(|v| v.to_time()).unwrap_or(default);
+                                let start = PrimitiveDateTime::new(start_date, start_time)
+                                    .assume_offset(std_offset);
+                                let end = PrimitiveDateTime::new(end_date, end_time)
+                                    .assume_offset(std_offset);
+                                if date_time >= &start && date_time <= &end {
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Fixed in latest commit.

File: src/parse_tz.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Should be resolved in latest commit.

File: src/parse_tz.rs [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+    Date,
+}
+
+pub enum Error<'a> {
+    Nom(ErrorKind),
+    UnknownName(&'a str),
+    Range(RangeError),
+}
+
+fn quoted_name(input: &str) -> IResult<&str, &str> {
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Fixed in latest commit.

File: src/parse_tz.rs [hunk 10]
```diff
@@ -0,0 +1,487 @@
...
+fn name(input: &str) -> IResult<&str, &str> {
+    alt((quoted_name, unquoted_name))(input)
+}
+
+fn time_component(input: &str) -> IResult<&str, u8> {
+    map_res(digit1, |v: &str| v.parse::<u8>())(input)
+}
+
+fn time_component_opt(input: &str) -> IResult<&str, u8> {
+    let (v, (_, v1)) = tuple((cchar(':'), time_component))(input)?;
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Fixed in latest commit.

File: src/parse_tz.rs [hunk 11]
```diff
@@ -0,0 +1,487 @@
...
+    alt((cchar('+'), cchar('-')))(input)
+}
+
+fn time(input: &str) -> IResult<&str, Time> {
+    let (input, (hh, mm, ss)) = tuple((
+        time_component,
+        opt(time_component_opt),
+        opt(time_component_opt),
+    ))(input)?;
+    Ok((input, Time { hh, mm, ss }))
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: In the mean time I've just added .map at the end which is more complicated but at least it compiles.

File: src/parse_tz.rs (hunk 1 above)
Comment (Yuri6037) [reply src/parse_tz.rs]: Fixed in latest commit.

File: src/parse_tz.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz.rs]: Fixed in latest commit with thiserror. Thanks for the reference to `thiserror` didn't know about it btw.

File: src/parse_tz/abstract.rs [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                            let start_date = v.start.0.to_date(utc.year()).map_err(|e| e.into())?;
+                            let end_date = v.end.0.to_date(utc.year()).map_err(|e| e.into())?;
+                            let start_time =
+                                v.start.1.as_ref().map(|v| v.to_time()).unwrap_or(default);
+                            let end_time = v.end.1.as_ref().map(|v| v.to_time()).unwrap_or(default);
+                            let start = PrimitiveDateTime::new(start_date, start_time)
+                                .assume_offset(std_offset);
+                            let end = PrimitiveDateTime::new(end_date, end_time)
+                                .assume_offset(dst_offset);
+                            Some(Rule { start, end })
```
Comment (MiSawa) [inline src/parse_tz/abstract.rs:159]: Hmmm actually I think this `start` and `end` has to reflect the `OffsetDateTime` given to `get_offset_utc` as the previous implementation, because the rule should have been applied to all years while we currently apply only on the current year.

**Review by MiSawa — COMMENTED** [review]
I've locally added a main function that compares the conversion result against what `TZ=... date` does. https://github.com/MiSawa/time-tz/blob/smoke-test/src/main.rs

File: src/parse_tz/abstract.rs (hunk 12 above)
Comment (Yuri6037) [reply src/parse_tz/abstract.rs:159]: Well that's the problem; if I get this calculation done on the OffsetDateTime then implementing TimeZone is impossible because of the fact that this conversion can return errors whereas TimeZone doesn't allow for errors. 

Adding error returns would be a big breaking change and wouldn't benefit the API that much because errors are only needed for the TZ parser path and not for for tzdb...

Comment (MiSawa) [reply src/parse_tz/abstract.rs:159]: Hmmmm....
Perhaps we can calculate 2*7 ((leap/non-leap)*(the weekday of Jan 1)) patterns of `start`/`end` here, and in `get_offset_utc`, select one of them according to the input `OffsetDateTime` and shift the year part?

Comment (Yuri6037) [reply src/parse_tz/abstract.rs:159]: time-rs doesn't allow shifting the year component.

Comment (MiSawa) [reply src/parse_tz/abstract.rs:159]: Hmmm, so then maybe store those 2 \* 7 `start`/`end` as `Duration`s from Jan 1st of that year?
With that, we can check if `((given datetime) - start).year() == (given datetime).year() && ((given datetime) - end).year() < (given datetime).year()`.

Comment (Yuri6037) [reply src/parse_tz/abstract.rs:159]: I think I made a big mistake: the TZ environment variable does not represent an actual timezone because it's not precise and generic. I think TZ is not designed to be interpreted the way we're trying to interpret it. It may be sometimes a full timezone (the :whatever syntax), however in the other form it doesn't match a generic date time it's designed to be used with the current system time which is different scope than TimeZone which assumes the timezone is generic and valid for any UTC date_time given as input to get_offset_utc...

Comment (MiSawa) [reply src/parse_tz/abstract.rs:159]: I don't fully understand the burden yet, probably require refreshing my brain, but yes, that API design should also work for my use case!

Comment (Yuri6037) [reply src/parse_tz/abstract.rs]: There we go, I've just finished designing the new API.

**Comment by Yuri6037** [comment]
> I've locally added a main function that compares the conversion result against what `TZ=... date` does. https://github.com/MiSawa/time-tz/blob/smoke-test/src/main.rs

I've tried adding this as unit tests but it doesn't seem cross platform. On Mac I get "date: settimeofday (timezone): Operation not permitted".

File: src/parse_tz/intermediate.rs [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+                // contains 28 days instead of 29 which matches the POSIX spec.
+                let date = time::Date::from_ordinal_date(2021, *n).map_err(BypassRustDefect::ComponentRange)?;
+                // Not sure if that will work in all cases though...
+                time::Date::from_calendar_date(year, date.month(), date.day()).map_err(BypassRustDefect::ComponentRange)
+            }
+            // ComponentRange errors should be prevented by is_valid_range.
+            Date::N(n) => time::Date::from_ordinal_date(year, *n + 1).map_err(BypassRustDefect::ComponentRange),
+            Date::M { m, n, d } => {
+                // One more hack: here w're trying to match Date::from_iso_week_date.
+                let month = match m {
```
[the remark this answers is not in the document]
Comment (Yuri6037) [reply src/parse_tz/intermediate.rs]: Fixed in latest commit.

File: src/posix_tz/abstract.rs [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+            let std = ExpandedMode {
+                name: std.name,
+                offset: std_offset,
+            };
+            let (dst, rule) = match dst {
+                None => (None, None),
+                Some(v) => {
+                    // If no offset is specified the POSIX standard defines +1 hour in standard
+                    // time as default.
+                    let offset = v.offset.map(|v| v.to_seconds()).unwrap_or(tmp + 3600);
```
Comment (MiSawa) [inline src/posix_tz/abstract.rs:140]: Oops my bad it's `-3600`, sorry.... (was confused with the negation of POSIX spec)

**Comment by MiSawa** [comment]
It works great! Thank you so much!

> > I've locally added a main function that compares the conversion result against what `TZ=... date` does. https://github.com/MiSawa/time-tz/blob/smoke-test/src/main.rs
> 
> I've tried adding this as unit tests but it doesn't seem cross platform. On Mac I get "date: settimeofday (timezone): Operation not permitted".

Yeah it's very much non-portable, as (I don't know much but) it calls an external command `date` which probably doesn't exists in Windows machine in the first place. And also the command line args of OSX is different from that of gnu ones. And this is one of the reasons why I didn't make it a `cargo test` (another reason is it's very slow). To run this on OSX, installing `coreutils` plus use `gdate` instead of `date` should work I believe.
The output currently differs for an edge case where the tz rules are very near to the year boundary like `ABC4DEF,J1/0,J365/23`, but practically that not going to happen, so I guess it's fine.

**Comment by Yuri6037** [comment]
> It works great! Thank you so much!
> 
> > > I've locally added a main function that compares the conversion result against what `TZ=... date` does. https://github.com/MiSawa/time-tz/blob/smoke-test/src/main.rs
> > 
> > 
> > I've tried adding this as unit tests but it doesn't seem cross platform. On Mac I get "date: settimeofday (timezone): Operation not permitted".
> 
> Yeah it's very much non-portable, as (I don't know much but) it calls an external command `date` which probably doesn't exists in Windows machine in the first place. And also the command line args of OSX is different from that of gnu ones. And this is one of the reasons why I didn't make it a `cargo test` (another reason is it's very slow). To run this on OSX, installing `coreutils` plus use `gdate` instead of `date` should work I believe. The output currently differs for an edge case where the tz rules are very near to the year boundary like `ABC4DEF,J1/0,J365/23`, but practically that not going to happen, so I guess it's fine.

Ok so I've found a way to make it run on Mac by replacing "date -d" with "date -r" (just looked at man date)... However I've noticed something strange: if the default DST offset is -3600 or std + 3600 it fails. I figured from the output that Mac date does something different it takes -STD + 3600 seconds as default instead.

Do you think that's normal or is it Apple not complying to POSIX?

**Comment by MiSawa** [comment]
> Ok so I've found a way to make it run on Mac by replacing "date -d" with "date -r" (just looked at man date)... However I've noticed something strange: if the default DST offset is -3600 or std + 3600 it fails. I figured from the output that Mac date does something different it takes -STD + 3600 seconds as default instead.
> 
> Do you think that's normal or is it Apple not complying to POSIX?

Negative `std` doesn't make sense to me... `std - 3600` does though. I mean, it should match the actual real-world daylight saving time that USA for example have. For example, if it's 9AM currently in STD and DST starts tomorrow, then 24 hour later should be 10AM, as to follow the fact that the sun raise earlier.

**Comment by Yuri6037** [comment]
> > Ok so I've found a way to make it run on Mac by replacing "date -d" with "date -r" (just looked at man date)... However I've noticed something strange: if the default DST offset is -3600 or std + 3600 it fails. I figured from the output that Mac date does something different it takes -STD + 3600 seconds as default instead.
> > Do you think that's normal or is it Apple not complying to POSIX?
> 
> Negative `std` doesn't make sense to me... `std - 3600` does though. I mean, it should match the actual real-world daylight saving time that USA for example have. For example, if it's 9AM currently in STD and DST starts tomorrow, then 24 hour later should be 10AM, as to follow the fact that the sun raise earlier.

I've checked again and it really chooses UTC - STD + 3600 as default offset.

**Comment by MiSawa** [comment]
Ahhh ooooops, I was really confused about what we refer with `std`, my bad. Yeah in my Linux box, it works if I changed
```
                     let offset = v.offset.map(|v| v.to_seconds()).unwrap_or(tmp + 3600); 
```
to
```
                     let offset = v.offset.map(|v| v.to_seconds()).unwrap_or(tmp - 3600); 
```
which means `-std+3600` (std here is the thing in input string).

File: src/posix_tz/abstract.rs (hunk 14 above)
Comment (Yuri6037) [reply src/posix_tz/abstract.rs]: Was actually -STD + 3600.

**Comment by Yuri6037** [comment]
Here that should fix it.

**Comment by MiSawa** [comment]
> Here that should fix it.

Confirmed, thank you!

**Review by MiSawa — APPROVED** [review]
